daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/ledgerinteraction/GrpcLedgerClient.scala

```diff
@@ -377,10 +377,20 @@ class GrpcLedgerClient(val grpcClient: LedgerClient, val applicationId: Applicat
       esf: ExecutionSequencerFactory,
       mat: Materializer,
   ): Future[List[User]] = {
-    // TODO https://github.com/digital-asset/daml/issues/12663 participant user management: Emulating no-pagination
-    grpcClient.userManagementClient.listUsers(pageToken = "", pageSize = 10000).map {
-      case (users, _) => users.toList
+    val pageSize = 100
+    def listWithPageToken(pageToken: String): Future[List[User]] = {
+      grpcClient.userManagementClient
+        .listUsers(pageToken = pageToken, pageSize = pageSize)
+        .flatMap { case (users, nextPageToken) =>
+          if (users.size < pageSize) Future.successful(users.toList)
+          else {
+            listWithPageToken(nextPageToken).map { more =>
+              users.toList ++ more
+            }
+          }
```

Comment:
I think this could result in an infinite loop: if the last page has `user.size == pageSize` you will use go in the `else` and the last page gives you the empty string as the next page token, making you loop back into the first page. I would recommend using `nextPageToken == ""` as the stopping condition as in the following example:

https://github.com/digital-asset/daml/blob/5f4dd96581eab4cac7cc9f9922d830172e416d4a/ledger-service/http-json/src/main/scala/com/digitalasset/http/Endpoints.scala#L495-L503

Reply:
That's not what I observed. You always get one final page with no users.

Reply:
Still seems like a reasonable change to make.You are probably right that it works with the current ledger behavior but I don’t think it’s a guarantee that is explicitly provided so better to not rely on it.

Reply:
We shouldn't have `nextPageToken == ""` be the only test, or else every call to `listAllUsers` will make a final pointless roundtrip over the `Grpc` interface. But we can add it as an extra check for safety's sake.

Reply:
> That's not what I observed. You always get one final page with no users.

@meiersi-da @pbatko-da Is this a quirk of some sort or an actual guarantee? I don't see this mentioned neither here

https://github.com/digital-asset/daml/blob/e3bae961ccce4fcb756b6d1af73bc7bcdde13536/ledger-api/grpc-definitions/com/daml/ledger/api/v1/admin/user_management_service.proto#L32-L33

nor here

https://github.com/digital-asset/daml/blob/e3bae961ccce4fcb756b6d1af73bc7bcdde13536/ledger-api/grpc-definitions/com/daml/ledger/api/v1/admin/user_management_service.proto#L137-L144

Reply:
(_quirk_ is not meant in a bad way, just in the sense of an implementation-specific detail that should not necessarily be relied upon)

Reply:
Just came across this comment from you @stefanobaghino-da , as I was behind in inbox processing. You're totally right in pointing out this *quirk*. As we discussed offline, it was addressed in https://github.com/digital-asset/daml/pull/12773